Fix Built Output Tool Window performance issues on projects with thousands of tasks

yuriy-solodkyy opened the pull request:
This PR address several problems in the build output tool window which significantly slow down the UI and overall build time especially for Android projects with many modules (above 200). This is extremely noticeable when all the Gradle tasks are up-to-date. 

(1) The dedicated thread updating the tree structure does not  keep up with the rate of events coming from Gradle because of acquiring the read lock for each event (and allocating a progress indicator. The  later problem is not addressed by this PR). 

(2) Running TreeUtil.expand() over the entire tree on each event (this operation allocates a lot of sync operations and becomes slower as the tree grows)

(3) CachingSimpleNode is "simple" and resets its cache of nodes on each change. It results in large arrays being re-allocated and a filter being applied to thousand of nodes on each event. 

See the commit descriptions for more specific details. 

@nskvortsov

nskvortsov commented on plugins/gradle/src/org/jetbrains/plugins/gradle/settings/GradleExtensionsSettings.java:
```diff
@@ -116,28 +116,28 @@ public void add(@NotNull String rootPath, @NotNull Map<String, GradleExtensions>
-          gradleTask.description = description.toString();
-          extensionsData.tasksMap.put(gradleTask.name, gradleTask);
-        }
+        //for (ExternalTask task : gradleExtensions.getTasks()) {
```
is this commenting out code intentional, or added to pull request by mistake?

yuriy-solodkyy replied:
This change has been moved to a separate PR and reverted there.

yuriy-solodkyy commented:
This is unrelated. I forgot to undo the change when moving the patches
between repos. It only helps me it testing as it is a different performance
bottleneck on a large projects. I'll reupload later today when get back to
my desk.

On Fri, Jan 24, 2020, 12:57 Nikita Skvortsov <notifications@github.com>
wrote:

> *@nskvortsov* commented on this pull request.
> ------------------------------
>
> In
> plugins/gradle/src/org/jetbrains/plugins/gradle/settings/GradleExtensionsSettings.java
> <https://github.com/JetBrains/intellij-community/pull/1260#discussion_r370619920>
> :
>
> > -
> -          StringBuilder description = new StringBuilder();
> -          if (task.getDescription() != null) {
> -            description.append(task.getDescription());
> -            if (task.getGroup() != null) {
> -              description.append("<p>");
> -            }
> -          }
> -          if (task.getGroup() != null) {
> -            description.append("<i>Task group: ").append(task.getGroup()).append("<i>");
> -          }
> -
> -          gradleTask.description = description.toString();
> -          extensionsData.tasksMap.put(gradleTask.name, gradleTask);
> -        }
> +        //for (ExternalTask task : gradleExtensions.getTasks()) {
>
> is this commenting out code intentional, or added to pull request by
> mistake?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JetBrains/intellij-community/pull/1260?email_source=notifications&email_token=AIEWSIQOMD5MIIN2HTLVUMTQ7LQR3A5CNFSM4JQTUGNKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCS6RHCQ#pullrequestreview-347935626>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIEWSIQJOXMM6UFC3KXE4UDQ7LQR3ANCNFSM4JQTUGNA>
> .
>

nskvortsov commented:
Can we also split this pull request into different parts?

Let's review changes in `Subscribe for fewer Gradle events`  in a separate pull request, as they are somewhat separate.
And we can concentrate on Tree related features here

yuriy-solodkyy commented:
> Can we also split this pull request into different parts?
> 
> Let's review changes in `Subscribe for fewer Gradle events` in a separate pull request, as they are somewhat separate.
I've split the PR and rebased this one on top of the new master. I'll upload the Gradle events part and update this thread with the link.
> And we can concentrate on Tree related features here

yuriy-solodkyy commented:
PR with Gradle related changes: https://github.com/JetBrains/intellij-community/pull/1296

nskvortsov commented on platform/lang-impl/src/com/intellij/build/ExecutionNode.java:
```diff
@@ -215,16 +238,21 @@ public void setFilter(@Nullable Predicate<ExecutionNode> filter) {
+  private void maybeReapplyFilter() {
+    if (myVisibleChildrenList != null) {
+      myVisibleChildrenList.clear();
+      myVisibleChildrenList.addAll(myChildrenList.stream().filter(it -> nodeIsVisible(it)).collect(Collectors.toList()));
```
Using `forEachOrdered` may save you building a large intermediate list (e.g., for thousands of nodes):
`myChildrenList.stream().filter(it -> nodeIsVisible(it)).forEachOrdered(visibleChildrenList::add)`

yuriy-solodkyy replied:
Done

nskvortsov commented on platform/lang-impl/src/com/intellij/build/BuildTreeConsoleView.java:
```diff
@@ -232,12 +244,17 @@ public void print(@NotNull String text, @NotNull ConsoleViewContentType contentT
     ExecutionNode parentNode = event.getParentId() == null ? null : nodesMap.get(event.getParentId());
     if (event instanceof MessageEvent) {
-      parentNode = createMessageParentNodes((MessageEvent)event, parentNode);
+      ExecutionNode node = parentNode;
```
Why do we need this variable here?

yuriy-solodkyy replied:
It used not to be ExecutionNode originally. I'll remove it.

nskvortsov replied:
thanks you

yuriy-solodkyy replied:
removed

yuriy-solodkyy commented on platform/lang-impl/src/com/intellij/build/BuildTreeConsoleView.java:
```diff
@@ -495,30 +523,25 @@ private Runnable addChildFailureNode(@NotNull ExecutionNode parentNode, @NotNull
+  private static void finishChildren(@NotNull SmartHashSet<ExecutionNode> structureChanged,
+                                     @NotNull ExecutionNode node,
+                                     @NotNull EventResult result) {
+    for (ExecutionNode child : node.getChildList()) {
```
This needs to be copied before being iterated over since `child.setResult(result)` may update the list. We learned this from canary builds. I'll update the PR a little later.

nskvortsov replied:
ok )

yuriy-solodkyy replied:
updated

yuriy-solodkyy replied:
done

nskvortsov commented:
We have reviewed the code, and it's good to go.
Let's clean up existing issues and I will merge the pull request.

yuriy-solodkyy commented:
I believe I addressed all your comments. Let me know if I missed anything. 
Thank you!

nskvortsov commented:
Thank you, will send it to our CI pipeline now

nskvortsov commented:
Please check `BuildTreeConsoleViewTest` - looks like some description for empty nodes has changed

yuriy-solodkyy commented:
There are in fact two unrelated failures in the test execution results:
1. The root node toString() returns "null" instead of "" - fixing. 
2. Auto-expansion: I changed it from 2 levels to just 1 level (visible root(s)). The reasons are: (a) in Gradle we don't have anything under tasks now (if we can still have warnings/errors the first one will expand anyway) and (b) the existing code does a lot of work to maintain auto-expansion and while my version does less it is still expensive. 
3. However, I'm looking at whether auto-expand can further be simplified. I'll get back to you soon. I need to test it.

yuriy-solodkyy commented:
1. I restored the old way to decide whether a node needs to be auto-expanded and fixed "null" problem. The test is green for me.  I was wrong, I don't think it affects performance now. Re-uploaded.

2. I cannot simplify it. I missed some details.

nskvortsov commented:
Merged manually, thank you!
